### PR TITLE
Initial import of Accommodation Booking Service

### PR DESCRIPTION
**Pull request by KIR4S3K**
- Initial codebase for Accommodation Booking Service  
- JWT auth (MANAGER & CUSTOMER)  
- CRUD APIs (accommodations, users, bookings, payments)  
- Stripe integration & Telegram notifications  
- Scheduled tasks (expire bookings/payments)  
- Liquibase migrations & Swagger/OpenAPI docs  
- Docker & Docker Compose setup

### CODE REVIEW

File: src/main/java/app/accommodationbookingservice/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                          UserService userService,
+                          PasswordEncoder encoder) {
+        this.authManager = authManager;
+        this.jwt = jwt;
+        this.userService = userService;
+        this.encoder = encoder;
+    }
+
+    @PostMapping("/register")
+    public ResponseEntity<User> register(@RequestBody User u) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/AuthController.java]: Don't name your variables like here `u` always use full and descriptive names like `user`.

File: src/main/java/app/accommodationbookingservice/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/auth")
+public class AuthController {
+    private final AuthenticationManager authManager;
+    private final JwtTokenProvider jwt;
+    private final UserService userService;
+    private final PasswordEncoder encoder;
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/AuthController.java]: You inject a PasswordEncoder but never use it on register—instead you pass plain text through `userService.register(u) `
. You should encode `u.setPassword(encoder.encode(u.getPassword()))` before saving.

File: src/main/java/app/accommodationbookingservice/controller/AuthController.java (hunk 1 above)
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/AuthController.java]: Exposing your JPA User directly is a security risk (you’ll leak fields like `password`). Swap to a `UserResponseDto`.

File: src/main/java/app/accommodationbookingservice/controller/UserController.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        User updated = userService.save(current);
+        return ResponseEntity.ok(userMapper.toDto(updated));
+    }
+
+    @Operation(summary = "Update user role (MANAGER only)")
+    @PutMapping("/{id}/role")
+    @PreAuthorize("hasRole('MANAGER')")
+    public ResponseEntity<UserDto> updateRole(
+            @PathVariable Long id,
+            @RequestBody String role) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/UserController.java]: You accept a raw `String role` in `updateRole`—consider wrapping it in a small `RoleUpdateDto `so you can apply `@NotBlank / @Pattern `to validate values.

File: src/main/java/app/accommodationbookingservice/controller/UserController.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public ResponseEntity<UserDto> me(Principal principal) {
+        User current = userService.findByEmail(principal.getName());
+        return ResponseEntity.ok(userMapper.toDto(current));
+    }
+
+    @Operation(summary = "Update current user profile")
+    @PutMapping("/me")
+    public ResponseEntity<UserDto> updateMe(@RequestBody UserDto dto, Principal principal) {
+        User current = userService.findByEmail(principal.getName());
+        current.setFirstName(dto.getFirstName());
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/UserController.java]: In `updateMe `you mutate fields manually; to keep controllers slim, you could delegate mapping to a `UserMapper.updateEntityFromDto(dto, user)`

File: src/main/java/app/accommodationbookingservice/controller/AccommodationController.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Accommodations", description = "Manage accommodation inventory")
+@RestController
+@RequestMapping("/accommodations")
+public class AccommodationController {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/AccommodationController.java:28]: You’re exposing `Accommodation `directly as both request and response. Better to introduce `AccommodationDto / CreateAccommodationDto` so you can validate inputs (`@Valid`) and avoid leaking fields like internal IDs or audit columns

File: src/main/java/app/accommodationbookingservice/controller/BookingController.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @PostMapping
+    public ResponseEntity<Booking> create(@RequestBody Booking b, Principal p) {
+        return ResponseEntity.ok(svc.create(b));
+    }
+
+    @GetMapping("/my")
+    public ResponseEntity<List<Booking>> my(Principal p) {
+        return ResponseEntity.ok(svc.findByUser(1L));
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/BookingController.java]: In `create `and `my`, you inject `Principal p `but ignore it—`my `hardcodes `findByUser(1L) `instead of `p.getName() → UserService.findByEmail → user.getId()`. This effectively always returns user 1’s bookings

File: src/main/java/app/accommodationbookingservice/controller/BookingController.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/bookings")
+public class BookingController {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/BookingController.java:27]: As with accommodations, don’t expose the Booking entity directly—use BookingDto and CreateBookingDto to validate dates (@Future) and ensure you never accept invalid date ranges.

File: src/main/java/app/accommodationbookingservice/controller/BookingController.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    private final BookingService svc;
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/BookingController.java]: Don't use name `svc`, during reading this code you have to go back to line 22 to see what service is it, it should be `BookingService` but without looking on line 22 you are not sure if this service is used. What in case if in the controller there are 2 services?

File: src/main/java/app/accommodationbookingservice/controller/PaymentController.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    @PostMapping
+    public ResponseEntity<Payment> createSession(
+            @RequestParam Long bookingId,
+            @RequestParam String success,
+            @RequestParam String cancel) {
+
+        return ResponseEntity.ok(
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/PaymentController.java]: `createSession `(should be 201 Created), `success `and `cancel `return plain strings—consider returning a PaymentResponseDto or MessageDto for consistency.

File: src/main/java/app/accommodationbookingservice/dto/AccommodationDto.java [hunk 10]
```diff
@@ -0,0 +1,4 @@
+package app.accommodationbookingservice.dto;
+
+public class AccommodationDto {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/dto/AccommodationDto.java:7]: It’s currently empty—no fields at all. Without fields, your mapper has nothing to populate—and clients won’t receive any data.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            NotificationService notificationService) {
+        this.repo = repo;
+        this.accommodationRepo = accommodationRepo;
+        this.paymentRepo = paymentRepo;
+        this.notificationService = notificationService;
+    }
+
+    @Override
+    public Booking create(Booking b) {
+        List<?> pendingPayments = paymentRepo.findPendingByUserId(b.getUserId());
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: `List<?>` is odd—your `PaymentRepository `should return` List<Payment>` so you can inspect fields or convert to DTO.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            throw new IllegalStateException(
+                    "You have pending payments; please complete them before booking.");
+        }
+
+        List<Booking> existing = repo.findByAccommodationIdAndStatusIn(
+                b.getAccommodationId(),
+                List.of(BookingStatus.PENDING, BookingStatus.CONFIRMED)
+        );
+        boolean overlaps = existing.stream().anyMatch(ex ->
+                !(b.getCheckOutDate().isBefore(ex.getCheckInDate())
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: Edge case: if one booking checks out on the same day another checks in, your logic marks that as overlapping (since `isBefore `is strict). If that’s not desired, adjust to `!b.getCheckOutDate().isBefore(ex.getCheckInDate().plusDays(1))`.

Add unit test to test here edge cases.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                        || b.getCheckInDate().isAfter(ex.getCheckOutDate()))
+        );
+        if (overlaps) {
+            throw new IllegalStateException(
+                    "Accommodation " + b.getAccommodationId()
+                            + " is already booked for these dates");
+        }
+
+        Accommodation acc = accommodationRepo.findById(b.getAccommodationId()).orElseThrow();
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: orElseThrow() with no message yields `NoSuchElementException`. Better to `.orElseThrow(() -> new EntityNotFoundException("Accommodation not found: " + id))` so clients get a 404.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        if (acc.getAvailability() <= 0) {
+            throw new IllegalStateException(
+                    "No availability for accommodation " + b.getAccommodationId());
+        }
+        acc.setAvailability(acc.getAvailability() - 1);
+        accommodationRepo.save(acc);
+
+        b.setStatus(BookingStatus.PENDING);
+        Booking saved = repo.save(b);
+        notificationService.notify("New booking created: " + saved.getId());
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: You may want to try/catch around the `notify `call so notification failures don’t cancel the booking.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+            PaymentRepository paymentRepo,
+            NotificationService notificationService) {
+        this.repo = repo;
+        this.accommodationRepo = accommodationRepo;
+        this.paymentRepo = paymentRepo;
+        this.notificationService = notificationService;
+    }
+
+    @Override
+    public Booking create(Booking b) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java:44]: Extract methods responsible for anything besides booking creation to separate methods. Currently this method is hard to maintain, hard to read and breaks SOLID. What about update method? We want similar validation there, so would you copy/paste this code to this method or rather reuse existing validation methods?

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        accommodationRepo.save(acc);
+
+        b.setStatus(BookingStatus.PENDING);
+        Booking saved = repo.save(b);
+        notificationService.notify("New booking created: " + saved.getId());
+        return saved;
+    }
+
+    @Override
+    public Booking update(Long id, Booking b) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java:98]: Missing validation:

1. Should re‐run the same overlap check if dates change.
2. Should validate `checkIn < checkOut.`
3. Should throw 404 if booking not found via a custom exception with `@ResponseStatus(HttpStatus.NOT_FOUND).`

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        Booking updated = repo.save(ex);
+        notificationService.notify("Booking updated: " + updated.getId());
+        return updated;
+    }
+
+    @Override
+    public void cancel(Long id) {
+        Booking ex = repo.findById(id).orElseThrow();
+        if (ex.getStatus() == BookingStatus.CANCELED) {
+            throw new IllegalStateException("Booking already canceled");
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: Use a custom `AlreadyCanceledException `(unchecked) annotated with` @ResponseStatus(HttpStatus.BAD_REQUEST) `instead of generic `IllegalStateException`.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+import com.stripe.model.checkout.Session;
+import com.stripe.param.checkout.SessionCreateParams;
+import java.math.BigDecimal;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java:26]: Class‐level `@Transactional` means every method is wrapped in a transaction. For read‐only methods (findById, findAll, etc.) you may want to annotate those with `@Transactional(readOnly = true)` or move them to a separate non‐transactional service interface.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+import app.accommodationbookingservice.repository.BookingRepository;
+import app.accommodationbookingservice.repository.PaymentRepository;
+import app.accommodationbookingservice.service.BookingService;
+import app.accommodationbookingservice.service.NotificationService;
+import java.util.List;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java:23]: Class‐level `@Transactional` means every method is wrapped in a transaction. For read‐only methods (findById, findAll, etc.) you may want to annotate those with `@Transactional(readOnly = true)` or move them to a separate non‐transactional service interface.

File: src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+import app.accommodationbookingservice.security.JwtTokenProvider;
+import app.accommodationbookingservice.service.UserService;
+import java.util.List;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java]: Class‐level `@Transactional` means every method is wrapped in a transaction. For read‐only methods (findById, findAll, etc.) you may want to annotate those with `@Transactional(readOnly = true)` or move them to a separate non‐transactional service interface.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    private String stripeKey;
+
+    public PaymentServiceImpl(PaymentRepository repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    public Payment createSession(Long bookingId, String successUrl, String cancelUrl) {
+        Stripe.apiKey = stripeKey;
+        BigDecimal amount = BigDecimal.valueOf(1000);
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: You hard‑code `amount = 1000`. According to your spec, you need to calculate the total price based on the booking dates and accommodation `dailyRate`.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
+public class PaymentServiceImpl implements PaymentService {
+    private final PaymentRepository repo;
+
+    @Value("${stripe.secret-key}")
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: `@Value` on a private field is fine, but constructor injection (passing the `stripeKey `into the constructor) makes the class easier to test (no reflection needed).

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+            Session session = Session.create(params);
+            Payment p = Payment.builder()
+                    .bookingId(bookingId)
+                    .sessionId(session.getId())
+                    .sessionUrl(session.getUrl())
+                    .status(PaymentStatus.PENDING)
+                    .amountToPay(amount)
+                    .build();
+            return repo.save(p);
+        } catch (Exception e) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: Catching `Exception `is too broad. Narrow it to `StripeException `(or `ApiException`) so you don’t mask unrelated errors.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+            throw new RuntimeException("Failed to create Stripe session", e);
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: Throw a custom unchecked exception (e.g. `PaymentException`) annotated with `@ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)` to give clients a 500 rather than a generic runtime error.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+
+    @Override
+    public Payment markAsPaid(String sessionId) {
+        Payment p = repo.findBySessionId(sessionId)
+                .orElseThrow(() -> new RuntimeException(
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: Instead of `RuntimeException`,` throw EntityNotFoundException("Payment not found: " + sessionId)`, annotated as `@ResponseStatus(HttpStatus.NOT_FOUND`).

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+                        "Payment with sessionId '" + sessionId + "' not found"));
+        p.setStatus(PaymentStatus.PAID);
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java:114]: If this endpoint is called twice, you’ll set `PAID `again. That’s harmless, but you might want to return a 400 if the payment is already marked as `PAID `or in a non‑PENDING state.

File: src/main/java/app/accommodationbookingservice/service/impl/TelegramNotificationServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TelegramNotificationServiceImpl implements NotificationService {
+
+    private static final Logger logger = LoggerFactory
+            .getLogger(TelegramNotificationServiceImpl.class);
+    private final OkHttpClient client = new OkHttpClient();
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/TelegramNotificationServiceImpl.java]: Inject OkHttpClient as a bean. Newing up new OkHttpClient() inside the service means you can’t share connection pools or configure timeouts globally.

File: src/main/java/app/accommodationbookingservice/service/impl/TelegramNotificationServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Value("${telegram.bot-token}")
+    private String token;
+
+    @Value("${telegram.chat-id}")
+    private String chatId;
+
+    @Override
+    public void notify(String message) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/TelegramNotificationServiceImpl.java:39]: Validate your configuration - If token or chatId is missing (blank), you currently build a malformed URL.

File: src/test/java/app/accommodationbookingservice/AccommodationBookingServiceApplicationTests.java [hunk 29]
```diff
@@ -0,0 +1,13 @@
+package app.accommodationbookingservice;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class AccommodationBookingServiceApplicationTests {
+
+    @Test
+    void contextLoads() {
```
Comment (slade13) [inline src/test/java/app/accommodationbookingservice/AccommodationBookingServiceApplicationTests.java]: Create test which checks actual things and remove this one.

File: src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public User register(User u) {
+        u.setPassword(encoder.encode(u.getPassword()));
+        u.setRole(UserRole.CUSTOMER);
+        return repo.save(u);
+    }
+
+    @Override
+    public User updateRole(Long id, String role) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java:51]: Be sure this method is only callable by a `MANAGER `via `@PreAuthorize("hasRole('MANAGER')") `at the controller or method level.

File: src/main/resources/application-prod.properties
Comment (slade13) [inline src/main/resources/application-prod.properties:1]: Introduce .env file to hold your secrets - as noted in the task description.

File: src/main/resources/application.properties [hunk 31]
```diff
@@ -0,0 +1,16 @@
+spring.profiles.active=dev
+
+# JWT
+jwt.secret=verySecretKeyChangeMe!
```
Comment (slade13) [inline src/main/resources/application.properties]: Never use a “verySecretKeyChangeMe!” in any committed properties. Default your` jwt.secret` from an env var (e.g. `jwt.secret=${JWT_SECRET:}`) and require it be set.

File: src/test/java/app/accommodationbookingservice/controller/AccommodationControllerTest.java [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+import java.math.BigDecimal;
+import java.util.List;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.ResponseEntity;
+
+class AccommodationControllerTest {
```
Comment (slade13) [inline src/test/java/app/accommodationbookingservice/controller/AccommodationControllerTest.java:25]: Test list_returnsEmpty when svc.findAll() returns empty list.
Test delete_notFound → 404 behavior when service throws.

**Review by slade13 — CHANGES_REQUESTED** [review]
Your project should consist of many commits/branches to show that you know how to work with git, currently you have only one commit which includes your whole project. Your code has some minor issues, please follow my comments on Github in order to fix that.

File: src/main/java/app/accommodationbookingservice/controller/AuthController.java [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+            UserService userService,
+            PasswordEncoder passwordEncoder) {
+        this.authenticationManager = authenticationManager;
+        this.jwtTokenProvider = jwtTokenProvider;
+        this.userService = userService;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    @PostMapping("/register")
+    public ResponseEntity<UserDto> register(@RequestBody RegisterRequestDto requestDto) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/controller/AuthController.java]: Missing `@Valid` annotation

File: src/main/java/app/accommodationbookingservice/dto/RegisterRequestDto.java [hunk 34]
```diff
@@ -0,0 +1,13 @@
+package app.accommodationbookingservice.dto;
+
+import app.accommodationbookingservice.model.enums.UserRole;
+import lombok.Data;
+
+@Data
+public class RegisterRequestDto {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/dto/RegisterRequestDto.java:11]: Add validation for this DTO.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,178 @@
...
+        this.accommodationRepo = accommodationRepo;
+        this.paymentRepo = paymentRepo;
+        this.notificationService = notificationService;
+    }
+
+    @Override
+    public Booking create(Booking b) {
+        List<Payment> pending = paymentRepo.findPendingByUserId(b.getUserId());
+        if (!pending.isEmpty()) {
+            throw new IllegalStateException(
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java]: You mix `ResponseStatusException(HttpStatus.BAD_REQUEST)` for validation errors with plain `IllegalStateException `for booking conflicts (which results in HTTP 500).
Use `ResponseStatusException `(or a custom exception subclass) consistently so clients receive the correct HTTP status and message. You can add tests which will validate those statuses.

File: src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java [hunk 36]
```diff
@@ -0,0 +1,178 @@
...
+                    "Accommodation " + b.getAccommodationId()
+                            + " is already booked for these dates"
+            );
+        }
+
+        Accommodation acc = accommodationRepo.findById(b.getAccommodationId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Accommodation not found: " + b.getAccommodationId()
+                ));
+        if (acc.getAvailability() <= 0) {
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java:75]: Concurrent requests can overshoot and book more than available units.
Apply a pessimistic lock (e.g. `@Lock(PESSIMISTIC_WRITE)` on your repository method) or perform the decrement in a single SQL update with `WHERE availability > 0`.

File: src/main/java/app/accommodationbookingservice/service/impl/AccommodationServiceImpl.java [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public Accommodation create(Accommodation a) {
+        return repo.save(a);
+    }
+
+    @Override
+    public Accommodation update(Long id, Accommodation a) {
+        Accommodation ex = repo.findById(id).orElseThrow();
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/AccommodationServiceImpl.java:26]: `orElseThrow()` without a message throws a `NoSuchElementException `with no context. Change it to something like this:
`.orElseThrow(() -> new EntityNotFoundException("Accommodation not found: " + id));`

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 38]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public Payment markAsPaid(String sessionId) {
+        Payment p = paymentRepo.findBySessionId(sessionId)
+                .orElseThrow(() -> new EntityNotFoundException("Payment not found: "
+                        + sessionId));
+        if (p.getStatus() != PaymentStatus.PENDING) {
+            throw new PaymentException("Cannot mark payment as paid in status: "
+                    + p.getStatus());
+        }
+        p.setStatus(PaymentStatus.PAID);
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java:114]: Marking a payment as `PAID `does not change the associated booking from `PENDING `to `CONFIRMED`. After `payment.setStatus(PAID)`, update `booking.setStatus(CONFIRMED)`.

File: src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java [hunk 39]
```diff
@@ -0,0 +1,127 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Accommodation not found: "
+                        + booking.getAccommodationId()));
+        long days = ChronoUnit.DAYS.between(booking.getCheckInDate(), booking
+                .getCheckOutDate());
+        BigDecimal amount = acc.getDailyRate().multiply(BigDecimal.valueOf(days));
+
+        SessionCreateParams.LineItem.PriceData priceData =
+                SessionCreateParams.LineItem.PriceData.builder()
+                        .setCurrency("usd")
+                        .setUnitAmount(amount.multiply(BigDecimal.valueOf(100)).longValue())
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/PaymentServiceImpl.java]: Converting `BigDecimal `to cents via `.longValue()` can silently truncate fractions. Use `.longValueExact()` or apply explicit rounding to avoid data loss.

File: src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java [hunk 40]
```diff
@@ -0,0 +1,68 @@
...
+            JwtTokenProvider jwt) {
+        this.repo = repo;
+        this.encoder = encoder;
+        this.jwt = jwt;
+    }
+
+    @Override
+    @Transactional
+    public User register(User u) {
+        u.setPassword(encoder.encode(u.getPassword()));
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java:44]: `register()` blindly saves, causing a `DataIntegrityViolationException `(500) on duplicates. Before encoding, do 
```
if (repo.existsByEmail(dto.getEmail())) 
    throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email already in use");
```

File: src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java [hunk 41]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    @Transactional
+    public User updateRole(Long id, String role) {
+        User u = repo.findById(id).orElseThrow(
+                () -> new UsernameNotFoundException("User not found: "
+                        + id)
+        );
+        u.setRole(UserRole.valueOf(role));
```
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java]: Unsafe `Role.valueOf() `can throw `IllegalArgumentException `for invalid input, resulting in a 500. Catch that exception and return 400 with a clear message listing valid roles.

File: src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java (hunk 40 above)
Comment (slade13) [inline src/main/java/app/accommodationbookingservice/service/impl/UserServiceImpl.java:44]: You accept any string as password. Enforce rules via DTO annotations (e.g. `@Size(min=8)`, `@Pattern`) for minimum strength

**Review by slade13 — APPROVED** [review]
Good job, I've added few more comments. Please review them and good luck!
